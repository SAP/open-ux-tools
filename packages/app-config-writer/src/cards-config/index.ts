import { join, basename } from 'path';
import type { Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import { getPreviewMiddleware, getIntentFromPreviewConfig, getCLIForPreview } from '../common/utils';
import type { MiddlewareConfig as PreviewConfig } from '@sap-ux/preview-middleware';
import type { ToolsLogger } from '@sap-ux/logger';
import { FileName, type Package, readUi5Yaml } from '@sap-ux/project-access';
import { updateMiddlewaresForPreview } from '../common/ui5-yaml';

/**
 * Updates the `ui5.yaml` file to add card generator path to preview middleware configuration.
 *
 * @param {Editor} fs - The `mem-fs-editor` instance used to read and write files.
 * @param {string} basePath - The path to the project root where the `ui5.yaml` file is located.
 * @param {string} [yamlPath] - Path to the `ui5.yaml` configuration file. Defaults to `ui5.yaml` in the project root.
 * @param {ToolsLogger} [logger] - logger instance for logging debug information.
 * @returns {Promise<void>} A promise that resolves when the `ui5.yaml` file has been successfully updated.
 */
async function updateMiddlewareConfigWithGeneratorPath(
    fs: Editor,
    basePath: string,
    yamlPath?: string,
    logger?: ToolsLogger
): Promise<void> {
    const ui5YamlFile = yamlPath ? basename(yamlPath) : FileName.Ui5Yaml;
    const ui5YamlConfig = await readUi5Yaml(basePath, ui5YamlFile, fs);
    const previewMiddleware = await getPreviewMiddleware(ui5YamlConfig);

    if (previewMiddleware) {
        previewMiddleware.configuration ??= {};

        const middlewareConfig = previewMiddleware.configuration as PreviewConfig;
        const cardGeneratorPath = middlewareConfig?.editors?.cardGenerator?.path ?? 'test/flpGeneratorSandbox.html';

        middlewareConfig.editors ??= {};
        middlewareConfig.editors.cardGenerator = {
            path: cardGeneratorPath
        };
        ui5YamlConfig.updateCustomMiddleware(previewMiddleware);
        fs.write(join(basePath, ui5YamlFile), ui5YamlConfig.toString());
        logger?.debug(`Updated preview middleware in ${ui5YamlFile}.`);
    }
}

/**
 * Updates the `package.json` file to include a script for starting the card generator.
 *
 * @param {string} basePath - The path to the project root where the `package.json` file is located.
 * @param {Editor} fs - The `mem-fs-editor` instance used to read and write files.
 * @param {string} [yamlPath] - Optional path to the `ui5.yaml` configuration file for retrieving middleware configurations.
 * @returns {Promise<void>} A promise that resolves when the `package.json` file has been successfully updated.
 */
async function updatePackageJson(basePath: string, fs: Editor, yamlPath?: string): Promise<void> {
    const packageJsonPath = join(basePath, 'package.json');
    if (!fs.exists(packageJsonPath)) {
        throw new Error('package.json not found');
    }

    const packageJson = (fs.readJSON(packageJsonPath) ?? {}) as Package;
    const previewMiddleware = await getPreviewMiddleware(undefined, basePath, yamlPath, fs);
    const intent = getIntentFromPreviewConfig(previewMiddleware?.configuration) ?? '#app-preview';
    const cardGeneratorPath =
        (previewMiddleware?.configuration as PreviewConfig)?.editors?.cardGenerator?.path ??
        'test/flpGeneratorSandbox.html';
    const cliForPreview = await getCLIForPreview(basePath, yamlPath ?? '', fs);

    packageJson.scripts ??= {};
    packageJson.scripts['start-cards-generator'] = `${cliForPreview} --open '${cardGeneratorPath}${intent}'`;
    fs.writeJSON(packageJsonPath, packageJson);
}

/**
 * Enables the card generator configuration for the given application.
 *
 * This function updates the `ui5.yaml` file to include the necessary middlewares for the card generator
 * and modifies the `package.json` file to add a script for starting the card generator.
 *
 * @param {string} basePath - The path to the project root.
 * @param {string} [yamlPath] - Optional path to the `ui5.yaml` configuration file.
 * @param {ToolsLogger} [logger] - Optional logger instance for logging messages.
 * @param {Editor} [fs] - Optional `mem-fs-editor` instance for file system operations. If not provided, a new instance will be created.
 * @returns {Promise<Editor>} A promise that resolves to the updated `mem-fs-editor` instance.
 */
export async function enableCardGeneratorConfig(
    basePath: string,
    yamlPath?: string,
    logger?: ToolsLogger,
    fs?: Editor
): Promise<Editor> {
    fs = fs ?? create(createStorage());
    await updateMiddlewaresForPreview(fs, basePath, yamlPath, logger);
    await updateMiddlewareConfigWithGeneratorPath(fs, basePath, yamlPath, logger);
    await updatePackageJson(basePath, fs, yamlPath);
    return fs;
}
