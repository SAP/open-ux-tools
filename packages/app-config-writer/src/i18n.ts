import type { TOptions } from 'i18next';
import i18next from 'i18next';
import smartLinksRes from './translations/smartlinks-config.json';
import navConfigRes from './translations/navigation-config.json';

export const SMART_LINKS_NS = 'app-config-writer:smartLinksConfig';
export const NAV_CONFIG_NS = 'app-config-writer:navConfig';

/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18n(): Promise<void> {
    await i18next.init({
        resources: {
            en: {
                [SMART_LINKS_NS]: smartLinksRes,
                [NAV_CONFIG_NS]: navConfigRes
            }
        },
        lng: 'en',
        fallbackLng: 'en',
        defaultNS: SMART_LINKS_NS, // Default since first to add translations
        ns: [SMART_LINKS_NS, NAV_CONFIG_NS]
    });
}

/**
 * Helper function facading the call to i18next.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    return i18next.t(key, options);
}

initI18n().catch(() => {
    // Ignore any errors since the write will still work
});
