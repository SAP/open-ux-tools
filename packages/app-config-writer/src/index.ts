export { getSmartLinksTargetFromPrompt } from './prompt';
export { generateSmartLinksConfig } from './smartlinks-config';
