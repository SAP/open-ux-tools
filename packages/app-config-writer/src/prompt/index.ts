export { getSmartLinksTargetFromPrompt, promptUserPass } from './smartlinks-config';
export { promptInboundNavigationConfig } from './navigation-config';
