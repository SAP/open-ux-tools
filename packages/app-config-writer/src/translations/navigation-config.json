{
    "error": {
        "sapAppNotDefined": "Manifest is missing required section 'sap.app'",
        "manifestNotFound": "Manifest not found at path: {{- path}}",
        "inboundExists": "Inbound definition already exists for key: {{inboundKey}}"
    },
    "prompt": {
        "validationWarning": {
            "inputRequired": "{{inputName}} input is required",
            "maxLength": "Maximum length: {{maxLength}} characters"
        },
        "message": {
            "semanticObject": "Semantic Object",
            "action": "Action",
            "title": "Title",
            "subtitle": "Subtitle (optional)",
            "overwrite": "Overwrite existing config"
        }
    }
}
