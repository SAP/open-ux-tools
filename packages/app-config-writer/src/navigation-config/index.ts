import type { Manifest, ManifestNamespace } from '@sap-ux/project-access';
import { FileName, getWebappPath } from '@sap-ux/project-access';
import type { Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import { mergeObjects } from '@sap-ux/ui5-config';
import { join } from 'path';
import { NAV_CONFIG_NS, t } from '../i18n';

/**
 * Adds a basic inbound navigation configuration to the application manifest.
 *
 * @param appRootPath app path
 * @param inboundConfig the inbound configuration to be written
 * @param inboundConfig.semanticObject semantic object
 * @param inboundConfig.action action
 * @param inboundConfig.title Represents a title; to make this property language dependent (recommended), use a key in double curly brackets '{{key}}' and add to i18n file
 * @param inboundConfig.subTitle optional, Represents a subtitle; to make this property language dependent (recommended), use a key in double curly brackets '{{key}}' and add to i18n file
 * @param overwrite overwrite existing config
 * @param fs file system reference
 * @returns file system reference
 */
export async function generateInboundNavigationConfig(
    appRootPath: string,
    { semanticObject, action, title, subTitle }: ManifestNamespace.Inbound[string],
    overwrite = false,
    fs?: Editor
): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    const { manifest, manifestPath } = await readManifest(appRootPath, fs);
    const inboundKey = `${semanticObject}-${action}`;

    if (!overwrite && manifest['sap.app'].crossNavigation?.inbounds[inboundKey]) {
        throw Error(t('error.inboundExists', { inboundKey, ns: NAV_CONFIG_NS }));
    }

    const inbound = {
        [inboundKey]: {
            semanticObject,
            action,
            title,
            subTitle,
            signature: {
                parameters: {},
                additionalParameters: 'allowed'
            }
        }
    };
    const crossNavigation: Manifest['sap.app']['crossNavigation'] = Object.assign(
        manifest['sap.app'].crossNavigation ?? {},
        {
            inbounds: mergeObjects(manifest['sap.app'].crossNavigation?.inbounds, inbound)
        }
    );

    fs.extendJSON(manifestPath, { 'sap.app': Object.assign(manifest['sap.app'], { crossNavigation }) });
    return fs;
}

/**
 * Validates the basic manifest structure and existence required for inbound navigation addition.
 *
 * @param appPath path to the application
 * @param fs Editor instance
 * @throws an error specifiying the validation failure
 * @returns the manifest object and manifest path
 */
export async function readManifest(appPath: string, fs: Editor): Promise<{ manifest: Manifest; manifestPath: string }> {
    const manifestPath = join(await getWebappPath(appPath), FileName.Manifest);
    const manifest = fs.readJSON(manifestPath) as unknown as Manifest;

    if (!manifest) {
        throw Error(t('error.manifestNotFound', { path: manifestPath, ns: NAV_CONFIG_NS }));
    }

    if (!manifest['sap.app']) {
        throw Error(t('error.sapAppNotDefined', { ns: NAV_CONFIG_NS }));
    }

    return {
        manifest,
        manifestPath
    };
}
