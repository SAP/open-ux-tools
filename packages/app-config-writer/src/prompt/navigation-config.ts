import type { PromptObject } from 'prompts';
import { prompt } from 'prompts';
import { NAV_CONFIG_NS, t } from '../i18n';
import type { ManifestNamespace } from '@sap-ux/project-access';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import type { Editor } from 'mem-fs-editor';
import { readManifest } from '../navigation-config';

/**
 * Prompt for inbound navigation configuration values.
 *
 * @param basePath the application root path
 * @returns inbound config or undefined if config prompting was aborted
 */
export async function promptInboundNavigationConfig(
    basePath: string
): Promise<{ config: Partial<ManifestNamespace.Inbound[string]> | undefined; fs: Editor }> {
    const fs = create(createStorage());
    const { manifest } = await readManifest(basePath, fs);

    const inbounds = manifest?.['sap.app']?.crossNavigation?.inbounds;
    let config = await prompt(getPrompts(Object.keys(inbounds ?? {})));
    // Exit if overwrite is false
    if (config?.overwrite === false) {
        config = undefined;
    }

    return {
        config,
        fs
    };
}

/**
 * Validates that text input does not have zero length and optionally is less than the specified maximum length.
 * Returns an end user message if validation fails.
 *
 * @param input the text input to validate
 * @param inputName the name of the input as seen by the user
 * @param maxLength optional, the maximum length of text to allow
 * @returns true, if all validation checks pass or a message explaining the validation failure
 */
function validateText(input: string, inputName: string, maxLength = 0): boolean | string {
    if (input?.trim().length === 0) {
        return t('prompt.validationWarning.inputRequired', {
            inputName,
            ns: NAV_CONFIG_NS
        });
    }

    if (maxLength && input.length > maxLength) {
        return t('prompt.validationWarning.maxLength', { maxLength, ns: NAV_CONFIG_NS });
    }
    return true;
}

/**
 * Get the prompts for inbound navigation configuration.
 *
 * @param inboundKeys inbound navigation keys already existing
 * @returns array of prompts
 */
function getPrompts(inboundKeys: string[]): PromptObject[] {
    const semanticObjectInputMsg = t('prompt.message.semanticObject', { ns: NAV_CONFIG_NS });
    const actionInputMsg = t('prompt.message.action', { ns: NAV_CONFIG_NS });
    const titleMsg = t('prompt.message.title', { ns: NAV_CONFIG_NS });

    return [
        {
            name: 'semanticObject',
            type: 'text',
            message: semanticObjectInputMsg,
            format: (val) => val?.trim(),
            validate: (val) => validateText(val, semanticObjectInputMsg, 30)
        },
        {
            name: 'action',
            type: 'text',
            message: actionInputMsg,
            format: (val) => val?.trim(),
            validate: (val) => validateText(val, actionInputMsg, 60)
        },
        {
            type: (prev, values) =>
                inboundKeys.indexOf(`${values.semanticObject}-${values.action}`) > -1 ? 'confirm' : false,
            name: 'overwrite',
            message: t('prompt.message.overwrite', { ns: NAV_CONFIG_NS }),
            initial: false
        },
        {
            name: 'title',
            type: (prev, values) => (values.overwrite !== false ? 'text' : false),
            message: titleMsg,
            format: (val) => val?.trim(),
            validate: (val) => validateText(val, titleMsg)
        },
        {
            name: 'subTitle',
            type: (prev, values) => (values.overwrite !== false ? 'text' : false),
            message: t('prompt.message.subtitle', { ns: NAV_CONFIG_NS }),
            format: (val) => val?.trim()
        }
    ];
}
