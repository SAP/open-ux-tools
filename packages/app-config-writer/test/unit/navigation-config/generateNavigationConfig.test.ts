import { join } from 'path';
import { generateInboundNavigationConfig } from '../../../src/navigation-config';

describe('Unit tests for navigation config generation', () => {
    const fixturePath = join(__dirname, '../../fixtures/navigation-config');

    test('generate inbound nav config entries', async () => {
        const appPathNoInbounds = join(fixturePath, '/ui5-app-no-inbounds');
        let fs = await generateInboundNavigationConfig(appPathNoInbounds, {
            semanticObject: 'semanticObject1',
            action: 'action1',
            title: 'title1',
            subTitle: 'subtitle1'
        });
        expect(fs.readJSON(join(appPathNoInbounds, 'webapp', 'manifest.json'))).toMatchSnapshot();

        // Ensure ui5 paraneters escaping is respected
        fs = await generateInboundNavigationConfig(appPathNoInbounds, {
            semanticObject: 'semanticObject1',
            action: 'action1',
            title: '{{title1}}',
            subTitle: '{{subtitle1}}'
        });
        expect(fs.readJSON(join(appPathNoInbounds, 'webapp', 'manifest.json'))).toMatchSnapshot();

        // Inbounds extended
        const appPathInbounds = join(fixturePath, '/ui5-app-inbounds');
        fs = await generateInboundNavigationConfig(appPathInbounds, {
            semanticObject: 'semanticObject2',
            action: 'action2',
            title: 'title2',
            subTitle: 'subtitle2'
        });
        expect(fs.readJSON(join(appPathInbounds, 'webapp', 'manifest.json'))).toMatchSnapshot();

        // Inbound warning exists
        const appPathInboundKeyExists = join(fixturePath, '/ui5-app-inbounds');
        await expect(() =>
            generateInboundNavigationConfig(appPathInboundKeyExists, {
                semanticObject: 'semanticObject1',
                action: 'action1',
                title: 'title2',
                subTitle: 'subtitle2'
            })
        ).rejects.toThrowErrorMatchingInlineSnapshot(
            `"Inbound definition already exists for key: semanticObject1-action1"`
        );

        // Inbounds override, should be replaced
        const appPathInboundsExist = join(fixturePath, '/ui5-app-inbounds');
        fs = await generateInboundNavigationConfig(
            appPathInboundsExist,
            { semanticObject: 'semanticObject1', action: 'action1', title: 'title3', subTitle: 'subtitle3' },
            true
        );
        expect(fs.readJSON(join(appPathInboundsExist, 'webapp', 'manifest.json'))).toMatchSnapshot();

        // Optional properties not provided
        fs = await generateInboundNavigationConfig(appPathNoInbounds, {
            semanticObject: 'semanticObject1',
            action: 'action1'
        });
        expect(fs.readJSON(join(appPathNoInbounds, 'webapp', 'manifest.json'))).toMatchSnapshot();
    });

    test('manifest not found', async () => {
        const appPathNoInbounds = join(fixturePath, '/not-existing-app');
        await expect(() =>
            generateInboundNavigationConfig(appPathNoInbounds, {
                semanticObject: 'semanticObject1',
                action: 'action1',
                title: 'title1',
                subTitle: 'subtitle1'
            })
        ).rejects.toThrowError(/^Manifest not found at path: /);
    });

    test('"sap.app" not defined', async () => {
        const appPathNoInbounds = join(fixturePath, '/sap-app-not-defined');
        await expect(() =>
            generateInboundNavigationConfig(appPathNoInbounds, {
                semanticObject: 'semanticObject1',
                action: 'action1',
                title: 'title1',
                subTitle: 'subtitle1'
            })
        ).rejects.toThrowErrorMatchingInlineSnapshot(`"Manifest is missing required section 'sap.app'"`);
    });
});
