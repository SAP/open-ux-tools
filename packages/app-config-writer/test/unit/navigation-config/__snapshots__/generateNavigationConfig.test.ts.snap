// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for navigation config generation generate inbound nav config entries 1`] = `
Object {
  "_version": "1.38.0",
  "sap.app": Object {
    "crossNavigation": Object {
      "inbounds": Object {
        "semanticObject1-action1": Object {
          "action": "action1",
          "semanticObject": "semanticObject1",
          "signature": Object {
            "additionalParameters": "allowed",
            "parameters": Object {},
          },
          "subTitle": "subtitle1",
          "title": "title1",
        },
      },
    },
    "id": "testNameSpace.alpv2",
    "type": "application",
  },
}
`;

exports[`Unit tests for navigation config generation generate inbound nav config entries 2`] = `
Object {
  "_version": "1.38.0",
  "sap.app": Object {
    "crossNavigation": Object {
      "inbounds": Object {
        "semanticObject1-action1": Object {
          "action": "action1",
          "semanticObject": "semanticObject1",
          "signature": Object {
            "additionalParameters": "allowed",
            "parameters": Object {},
          },
          "subTitle": "{{subtitle1}}",
          "title": "{{title1}}",
        },
      },
    },
    "id": "testNameSpace.alpv2",
    "type": "application",
  },
}
`;

exports[`Unit tests for navigation config generation generate inbound nav config entries 3`] = `
Object {
  "_version": "1.38.0",
  "sap.app": Object {
    "applicationVersion": Object {
      "version": "0.0.1",
    },
    "crossNavigation": Object {
      "inbounds": Object {
        "semanticObject1-action1": Object {
          "action": "action1",
          "semanticObject": "semanticObject1",
          "subTitle": "subtitle1",
          "title": "title1",
        },
        "semanticObject2-action2": Object {
          "action": "action2",
          "semanticObject": "semanticObject2",
          "signature": Object {
            "additionalParameters": "allowed",
            "parameters": Object {},
          },
          "subTitle": "subtitle2",
          "title": "title2",
        },
      },
    },
    "description": "{{appDescription}}",
    "i18n": "i18n/i18n.properties",
    "id": "testNameSpace.alpv2",
    "title": "{{appTitle}}",
    "type": "application",
  },
}
`;

exports[`Unit tests for navigation config generation generate inbound nav config entries 5`] = `
Object {
  "_version": "1.38.0",
  "sap.app": Object {
    "applicationVersion": Object {
      "version": "0.0.1",
    },
    "crossNavigation": Object {
      "inbounds": Object {
        "semanticObject1-action1": Object {
          "action": "action1",
          "semanticObject": "semanticObject1",
          "signature": Object {
            "additionalParameters": "allowed",
            "parameters": Object {},
          },
          "subTitle": "subtitle3",
          "title": "title3",
        },
      },
    },
    "description": "{{appDescription}}",
    "i18n": "i18n/i18n.properties",
    "id": "testNameSpace.alpv2",
    "title": "{{appTitle}}",
    "type": "application",
  },
}
`;
